### ISSUE TITLE
Unable to use DrawingBrush or VisualBrush as OpacityMask

### ISSUE THREAD

**Issue by emmauss** [AvaloniaUI/Avalonia#11388]
**Describe the bug**
When attempting to use a DrawingBrush or VisualBrush as Opacity, the following exception is thrown,
```
System.InvalidCastException
  HResult=0x80004002
  Message=Unable to cast object of type 'Avalonia.Media.VisualBrush' to type 'Avalonia.Media.IImmutableBrush'.
  Source=Avalonia.Base
  StackTrace:
   at Avalonia.Media.BrushExtensions.ToImmutable(IBrush brush) in /_/src/Avalonia.Base/Media/BrushExtensions.cs:line 23
   at Avalonia.Rendering.Composition.CompositingRenderer.UpdateCore() in /_/src/Avalonia.Base/Rendering/Composition/CompositingRenderer.cs:line 258
   at Avalonia.Rendering.Composition.CompositingRenderer.Update() in /_/src/Avalonia.Base/Rendering/Composition/CompositingRenderer.cs:line 314
   at Avalonia.Rendering.Composition.Compositor.CommitCore() in /_/src/Avalonia.Base/Rendering/Composition/Compositor.cs:line 172
   at Avalonia.Rendering.Composition.Compositor.Commit() in /_/src/Avalonia.Base/Rendering/Composition/Compositor.cs:line 95
   at Avalonia.Rendering.Composition.Compositor.<.ctor>b__27_0() in /_/src/Avalonia.Base/Rendering/Composition/Compositor.cs:line 61
   at Avalonia.Threading.DispatcherOperation.InvokeCore() in /_/src/Avalonia.Base/Threading/DispatcherOperation.cs:line 275

  This exception was originally thrown at this call stack:
    Avalonia.Media.BrushExtensions.ToImmutable(Avalonia.Media.IBrush) in BrushExtensions.cs
    Avalonia.Rendering.Composition.CompositingRenderer.UpdateCore() in CompositingRenderer.cs
    Avalonia.Rendering.Composition.CompositingRenderer.Update() in CompositingRenderer.cs
    Avalonia.Rendering.Composition.Compositor.CommitCore() in Compositor.cs
    Avalonia.Rendering.Composition.Compositor.Commit() in Compositor.cs
    Avalonia.Rendering.Composition.Compositor..ctor.AnonymousMethod__27_0() in Compositor.cs
    Avalonia.Threading.DispatcherOperation.InvokeCore() in DispatcherOperation.cs
```
Currently only color brushes, like SolidColorBrushes and GradiantBrush types can be used as OpacityMasks.

**To Reproduce**

1. Create a new Avalonia Project
2. Change the contents of MainView.axaml to this
 ```
<Grid Background="Red">
          <Grid.OpacityMask>
              <VisualBrush>
                  <VisualBrush.Visual>
                      <Grid Width="100" Height="100"/>
                  </VisualBrush.Visual>
              </VisualBrush>
          </Grid.OpacityMask>
          
</Grid>
```
3. Run it.

**Expected behavior**
Behavior should be similar to WPF's, with the mask having an effect.

**Screenshots**

**Desktop (please complete the following information):**

- OS: [e.g. Windows, Mac, Linux (State distribution)]
- Version [e.g. 0.10.0-rc1 or 0.9.12]

**Additional context**

**Comment by emmauss** [comment]
@kekekeks

**Comment by wenndemann** [comment]
Any news to this topic?

**Comment by wenndemann** [comment]
Too bad. Is there any workaround to specify a OpacityMask with a specific geometry?

**Comment by timunie** [comment]
If you need priority you can either try to file a PR or request custom development. I have no idea of a workaround but others may have.

**Comment by Gillibald** [comment]
No work towards this has been done yet. The issue is that a sub-scene has to be recorded in an immutable way so the compositor isn't facing any threading issues.

**Comment by StefanKoell** [comment]
Just to let you know, I'm facing the same issue.

**Comment by StefanKoell** [comment]
Here's what I'm trying to achieve using the Scrollviewer's opacity mask

```
                        <ScrollViewer.OpacityMask>
                            <VisualBrush>
                                <VisualBrush.Visual>
                                    <Grid RowDefinitions="50,*,50">
                                        <Border Grid.Row="0">
                                            <Border.BorderBrush>
                                                <LinearGradientBrush StartPoint="0%,0%" EndPoint="0%,100%">
                                                    <GradientStop Offset="0" Color="Transparent" />
                                                    <GradientStop Offset="1" Color="Black" />
                                                </LinearGradientBrush>
                                            </Border.BorderBrush>
                                        </Border>
                                        <Border Grid.Row="1" Background="Black"/>
                                        <Border Grid.Row="2">
                                            <Border.BorderBrush>
                                                <LinearGradientBrush StartPoint="0%,0%" EndPoint="0%,100%">
                                                    <GradientStop Offset="0" Color="Black" />
                                                    <GradientStop Offset="1" Color="Transparent" />
                                                </LinearGradientBrush>
                                            </Border.BorderBrush>
                                        </Border>
                                    </Grid>
                                </VisualBrush.Visual>
                            </VisualBrush>
                        </ScrollViewer.OpacityMask>
```

**Comment by timunie** [comment]
@StefanKoell workaround for your case could be a converter

```cs
public static class MyConv
{
    public static FuncValueConverter<double, double> GradientStopConverter { get; } =
        new FuncValueConverter<double, double>(x => 50.0 / x);
} 
```

```xml
<ScrollViewer x:Name="root">
    <ScrollViewer.OpacityMask>
        <LinearGradientBrush StartPoint="0%,0%" EndPoint="0%,100%">
            <GradientStop Offset="0"  Color="Transparent" />
            <GradientStop Offset="{Binding #root.Bounds.Height, Converter={x:Static app:MyConv.GradientStopConverter}}" Color="Black" />
        </LinearGradientBrush>
    </ScrollViewer.OpacityMask>
```

**Comment by Quertie** [comment]
Hello,

I am currently using Avalonia 11.1.4 and I believe the fix was made in 11.1.2, but I am still getting this error when using a VisualBrush as an OpacityMask.
I've migrated from an earlier version for this exact reason and deleted the package files and run `dotnet restore` just in case, but the problem persists.
Am I doing something wrong or has the fix not gone through ?

Here's my code for reference.
```XML
<Image Source="/Assets/image.png">
    <Image.OpacityMask>
        <VisualBrush>
            <VisualBrush.Visual>
                <Border Background="White" Height="12" Width="8" CornerRadius="4" />
            </VisualBrush.Visual>
        </VisualBrush>
    </Image.OpacityMask>
</Image>
```

Thank you for your time :)

**Comment by rabbitism** [comment]
> Hello,
> 
> @Gillibald Thanks for the fix !
> 
> I am currently using Avalonia 11.1.4 and I believe the fix was made in 11.1.2, but I am still getting this error when using a VisualBrush as an OpacityMask.
> I've migrated from an earlier version for this exact reason and deleted the package files and run `dotnet restore` just in case, but the problem 

This fix is not backported. It should be available in 11.2.x

**Comment by Quertie** [comment]
Oh I'm so sorry I guess I read the versions wrong.
Thanks for your prompt response ! 😊
